contracts/FlashRollover.sol

```diff
@@ -0,0 +1,344 @@
+
+import "./external/interfaces/ILendingPool.sol";
+
+import "./interfaces/IFlashRollover.sol";
+import "./interfaces/ILoanCore.sol";
+import "./interfaces/IOriginationController.sol";
+import "./interfaces/IRepaymentController.sol";
+import "./interfaces/IAssetWrapper.sol";
+import "./interfaces/IFeeController.sol";
+
+contract FlashRollover is IFlashRollover {
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /**
+     * Holds parameters passed through flash loan
+     * control flow that dictate terms of the new loan.
+     * Contains a signature by lender for same terms.
+     * isLegacy determines which loanCore to look for the
+     * old loan in.
+     */
+    struct OperationData {
+        bool isLegacy;
+        uint256 loanId;
+        LoanLibrary.LoanTerms newLoanTerms;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    /**
+     * Defines the contracts that should be used for a
+     * flash loan operation. May change based on if the
+     * old loan is on the current loanCore or legacy (in
+     * which case it requires migration).
+     */
+    struct OperationContracts {
+        ILoanCore loanCore;
+        IERC721 borrowerNote;
+        IERC721 lenderNote;
+        IFeeController feeController;
+        IERC721 assetWrapper;
+        IRepaymentController repaymentController;
+        IOriginationController originationController;
+        ILoanCore newLoanLoanCore;
+        IERC721 newLoanBorrowerNote;
+    }
+
+    /* solhint-disable var-name-mixedcase */
+    // AAVE Contracts
+    // Variable names are in upper case to fulfill IFlashLoanReceiver interface
+    ILendingPoolAddressesProvider public immutable override ADDRESSES_PROVIDER;
+    ILendingPool public immutable override LENDING_POOL;
+    /* solhint-enable var-name-mixedcase */
+
+    // Pawn.fi Contracts
+    ILoanCore public immutable loanCore;
+    ILoanCore public immutable legacyLoanCore;
+    IOriginationController public immutable originationController;
+    IRepaymentController public immutable legacyRepaymentController;
+    IRepaymentController public immutable repaymentController;
+    IERC721 public immutable borrowerNote;
+    IERC721 public immutable lenderNote;
+    IERC721 public immutable legacyBorrowerNote;
+    IERC721 public immutable legacyLenderNote;
+    IERC721 public immutable assetWrapper;
+    IFeeController public immutable feeController;
+
+    constructor(
+        ILendingPoolAddressesProvider _addressesProvider,
+        ILoanCore _loanCore,
+        ILoanCore _legacyLoanCore,
+        IOriginationController _originationController,
+        IRepaymentController _repaymentController,
+        IRepaymentController _legacyRepaymentController,
+        IERC721 _borrowerNote,
+        IERC721 _legacyBorrowerNote,
+        IERC721 _lenderNote,
+        IERC721 _legacyLenderNote,
+        IERC721 _assetWrapper,
+        IFeeController _feeController
```

Comment:
Would it make sense to just take `loanCoreFrom` and `loanCoreTo` in `rolloverLoan` and fetch all of these other addresses from the given loan core? That would be nice because the same contract could be used for future loan core instances that adhere to the same interface, and it would simplify a lot of the `isLegacy` logic and stuff... Tradeoff is runtime gas (though it might not even be that large of a tradeoff due to the number of `sload`s we have to do to get contracts anyways

Reply:
Ok really good point that we can get everything from `LoanCore`. I hated this (and had to deal with a lot of 'too many local variable' issues on my first writing). Will definitely do this but may change the PR pretty significantly

Reply:
Doing it this way added about 8k gas for a rollover, which I think is acceptable for the extra flexibility.

One thing to note is that you cannot get the controllers from loan core - the instance variables exist but they are never set. They will need to be passed into `rolloverLoan` as well